cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/trigger/DefaultTriggerFactory.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright © 2017 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.runtime.schedule.trigger;
+
+import co.cask.cdap.api.ProgramStatus;
+import co.cask.cdap.api.app.ProgramType;
+import co.cask.cdap.api.schedule.Trigger;
+import co.cask.cdap.api.schedule.TriggerFactory;
+import co.cask.cdap.proto.id.ApplicationId;
+import co.cask.cdap.proto.id.DatasetId;
+import co.cask.cdap.proto.id.NamespaceId;
+import co.cask.cdap.proto.id.ProgramId;
+
+/**
+ * The default implementation of {@link TriggerFactory}
+ */
+public class DefaultTriggerFactory implements TriggerFactory {
+
+  private final NamespaceId namespaceId;
+
+  public DefaultTriggerFactory(NamespaceId namespaceId) {
+    this.namespaceId = namespaceId;
+  }
+
+  @Override
+  public Trigger or(Trigger... triggers) {
+    return new OrTriggerBuilder(triggers);
+  }
+
+  @Override
+  public Trigger and(Trigger... triggers) {
+    return new AndTriggerBuilder(triggers);
+  }
+
+  @Override
+  public Trigger byTime(String cronExpression) {
+    return new TimeTrigger(cronExpression);
+  }
+
+  @Override
+  public Trigger onPartitions(String datasetName, int numPartitions) {
+    return new PartitionTrigger(namespaceId.dataset(datasetName), numPartitions);
+  }
+
+  @Override
+  public Trigger onPartitions(String datasetNamespace, String datasetName, int numPartitions) {
+    return new PartitionTrigger(new DatasetId(datasetNamespace, datasetName), numPartitions);
+  }
+
+  @Override
+  public Trigger onProgramStatus(String programNamespace, String application, String appVersion,
```

Comment:
Originally, I had named it `programNamespace` because `namespace` would conflict with the namespace defined in `DefaultScheduleBuilder`. Now that it's named `namespaceId`, there's no reason not to just use `namespace` as the variable name here.